File: 0222-Count Complete Tree Nodes/0222-Count Complete Tree Nodes.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int countNodes(TreeNode* root) {
        int h = height(root);
        int count = 0;
        while (root)
        {
            int rh = height(root->right);
            if (rh == h - 1)
            {
                count += (1 << h);
                root = root->right;
            }
            else
            {
                count += (1 << (h - 1));
                root = root->left;
            }
            --h;
        }
        
        return count;        
    }
    
private:
    int height(TreeNode* root)
    {
        int h = -1;
        while (root)
        {
            ++h;
            root = root->left;
        }
        return h;
    }
};

File: 0201-0300/0222-Count Complete Tree Nodes/0222-Count Complete Tree Nodes.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def countNodes(self, root: TreeNode) -> int:
        def height(root):
            h = -1
            while root:
                root = root.left
                h += 1
            return h
        
        h = height(root)
        count = 0
        while h >= 0:
            rh = height(root.right)
            if rh == h - 1:
                count += (1 << h)
                root = root.right
            else:
                count += (1 << (h - 1))
                root = root.left
            h -= 1
        return count
